KeychainMinder/main.m

```diff
@@ -77,40 +107,72 @@ OSStatus MechanismDestroy(AuthorizationMechanismRef inMechanism) {
 OSStatus MechanismInvoke(AuthorizationMechanismRef inMechanism) {
   MechanismRecord *mechanism = (MechanismRecord *)inMechanism;
   @autoreleasepool {
+    uid_t uid = GetUIDFromContex(mechanism);
+    gid_t gid = GetGIDFromContex(mechanism);
+
+    // Make sure this is not a hidden user.
+    if (uid < 501) {
+      return AllowLogin(mechanism);
+    }
+
     NSString *username = GetStringFromContext(mechanism, kAuthorizationEnvironmentUsername);
     NSString *password = GetStringFromContext(mechanism, kAuthorizationEnvironmentPassword);
+    NSString *sesOwner = GetStringFromHint(mechanism, kAuthorizationEnvironmentSuggestedUser);
+
+    // Make sure we have username and password data.
+    if (!username && !password) {
+      return AllowLogin(mechanism);
+    }
+
+    // Make sure the auth user is the sesion owner.
+    if (![username isEqualToString:sesOwner]) {
+      return AllowLogin(mechanism);
+    }
 
-    if (username && password && ![username hasPrefix:@"_"]) {
-      // Get the user's UID/GID from their username
-      struct passwd *pw = getpwnam([username UTF8String]);
-      uid_t uid = pw->pw_uid;
-      uid_t gid = pw->pw_gid;
-      endpwent();
-
-      // Switch EUID/EGID to the target user so SecKeychain* knows who to affect, validate
-      // the login keychain password, then switch back to the previous user.
-      BOOL passwordValid = YES;
-      if (setegid(gid) == 0 && seteuid(uid) == 0) {
-        SecKeychainSetUserInteractionAllowed(NO);
-        passwordValid = ValidateLoginKeychainPassword(password);
-        (void)seteuid(getuid());  // purposefully ignore return values when
-        (void)setegid(getgid());  // resetting back the effective uid/gid.
-      }
-
-      // Remove the current user, so they aren't duplicated in a second if
-      // the password wasn't valid.
-      NSMutableArray *users = GetUsers();
-      [users removeObject:username];
-
-      if (!passwordValid) {
-        [users addObject:username];
-      }
-      SetUsers(users);
+    BOOL keychainPasswordValid = YES;
+
+    // Switch the per thread EUID/EGID to the target user so SecKeychain* knows who to affect,
+    // validate the login keychain password, then switch back to the previous user.
+    // Using pthread as to not disrupt all of authorizationhost.
+    if (pthread_setugid_np(uid, gid) == 0) {
```

Comment:
I'd flip this logic, if it's `!= 0`: return, then the other stuff doesn't need to be indented.